### ISSUE TITLE
NUMERIC bind values passed to SQL Server VALUES constructor may lose scale if not explicit DataType is supplied

### ISSUE THREAD

**Issue by assylias** [jOOQ/jOOQ#17401]
### Expected behavior

The merge statement should produce the same results when run by jOOQ or run manually in Azure Data Studio and the numeric(19, 14) values should not get rounded to integers.

Note: using `dsl.settings().statementType = StatementType.STATIC_STATEMENT` solves the issue.

### Actual behavior

I run a merge that contains BigDecimal numbers (numeric(19, 14)). The detailed log is below:

```
2024-09-30 15:21:37,696 DEBUG [org.jooq.tools.LoggerListener] -> with bind values      :
		merge into [instrument].[Tranched] using (values (50893, 50811, 0, 0.024, 0.056), (50894, 50811, 0, 0.012, 0.028), (50895, 50819, 0.0423, 0.16, 1), (50896, 50819, 0, 0.08, 0.16), (50897, 50811, 0.0565, 0, 0.015), (50900, 50750, 0, 0.032, 0.064), (50901, 50750, 0.035, 0.016, 0.032)) [input] ([InstrumentId], [SeriesId], [FixedCoupon], [AttachmentPoint], [DetachmentPoint]) on [input].[InstrumentId] = [instrument].[Tranched].[InstrumentId] 
		when matched then update set [instrument].[Tranched].[InstrumentId] = [input].[InstrumentId], [instrument].[Tranched].[SeriesId] = [input].[SeriesId], [instrument].[Tranched].[FixedCoupon] = [input].[FixedCoupon], [instrument].[Tranched].[AttachmentPoint] = [input].[AttachmentPoint], [instrument].[Tranched].[DetachmentPoint] = [input].[DetachmentPoint]
		when not matched then insert ([InstrumentId], [SeriesId], [FixedCoupon], [AttachmentPoint], [DetachmentPoint]) values ([input].[InstrumentId], [input].[SeriesId], [input].[FixedCoupon], [input].[AttachmentPoint], [input].[DetachmentPoint]);
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 1       : 50893 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 2       : 50811 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 3       : 0 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 4       : 0.024 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 5       : 0.056 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 6       : 50894 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 7       : 50811 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 8       : 0 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 9       : 0.012 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 10      : 0.028 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 11      : 50895 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 12      : 50819 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 13      : 0.0423 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 14      : 0.16 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,697 TRACE [org.jooq.impl.DefaultBinding] Binding variable 15      : 1 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 16      : 50896 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 17      : 50819 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 18      : 0 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 19      : 0.08 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 20      : 0.16 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 21      : 50897 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 22      : 50811 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 23      : 0.0565 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 24      : 0 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 25      : 0.015 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 26      : 50900 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 27      : 50750 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 28      : 0 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 29      : 0.032 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 30      : 0.064 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 31      : 50901 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 32      : 50750 (integer /* java.lang.Integer */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 33      : 0.035 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 34      : 0.016 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,698 TRACE [org.jooq.impl.DefaultBinding] Binding variable 35      : 0.032 (numeric(19, 14) /* java.math.BigDecimal */)
2024-09-30 15:21:37,701 DEBUG [org.jooq.tools.LoggerListener] Affected row(s)          : 7
```

When jOOQ runs the merge query, the numeric(19,14) values are all rounded, meaning that they are all 0 except the 1 in record 50895.
When I run the exact same query (copy paste) from Azure Data Studio, the table gets updated as expected.

Looking at the logs in SQL server, I see that some of the values are multiplied by 1000 (e.g. 0.024 becoms 24 in the record 50893) but some are not (e.g. 0.16 in record 50895). But I'm not very familiar with this so it may be normal.

```
N'merge into [instrument].[Tranched] using (values ( @P0 ,  @P1 ,  @P2 ,  @P3 ,  @P4 ), ( @P5 ,  @P6 ,  @P7 ,  @P8 ,  @P9 ), ( @P10 ,  @P11 ,  @P12 ,  @P13 ,  @P14 ), ( @P15 ,  @P16 ,  @P17 ,  @P18 ,  @P19 ), ( @P20 ,  @P21 ,  @P22 ,  @P23 ,  @P24 ), ( @P25 ,  @P26 ,  @P27 ,  @P28 ,  @P29 ), ( @P30 ,  @P31 ,  @P32 ,  @P33 ,  @P34 )) [input] ([InstrumentId], [SeriesId], [FixedCoupon], [AttachmentPoint], [DetachmentPoint])
on [input].[InstrumentId] = [instrument].[Tranched].[InstrumentId]
when matched then update set [instrument].[Tranched].[InstrumentId] = [input].[InstrumentId], [instrument].[Tranched].[SeriesId] = [input].[SeriesId], [instrument].[Tranched].[FixedCoupon] = [input].[FixedCoupon], [instrument].[Tranched].[AttachmentPoint] = [input].[AttachmentPoint], [instrument].[Tranched].[DetachmentPoint] = [input].[DetachmentPoint]
when not matched then insert ([InstrumentId], [SeriesId], [FixedCoupon], [AttachmentPoint], [DetachmentPoint]) values ([input].[InstrumentId], [input].[SeriesId], [input].[FixedCoupon], [input].[AttachmentPoint], [input].[DetachmentPoint]);'
,
N'@P0 int,@P1 int,@P2 decimal(38,0),@P3 decimal(38,3),@P4 decimal(38,3),@P5 int,@P6 int,@P7 decimal(38,0),@P8 decimal(38,3),@P9 decimal(38,3),@P10 int,@P11 int,@P12 decimal(38,4),@P13 decimal(38,2),@P14 decimal(38,0),@P15 int,@P16 int,@P17 decimal(38,0),@P18 decimal(38,2),@P19 decimal(38,2),@P20 int,@P21 int,@P22 decimal(38,4),@P23 decimal(38,0),@P24 decimal(38,3),@P25 int,@P26 int,@P27 decimal(38,0),@P28 decimal(38,3),@P29 decimal(38,3),@P30 int,@P31 int,@P32 decimal(38,3),@P33 decimal(38,3),@P34 decimal(38,3)'
,
50893,50811,0,24,56,50894,50811,0,12,28,50895,50819,423,0.16,1,50896,50819,0,8,0.16,50897,50811,565,0,15,50900,50750,0,32,64,50901,50750,35,16,32
```

### Table definition
```
CREATE TABLE [instrument].[Tranched] (
    [InstrumentId]     INT              NOT NULL,
    [SeriesId]         INT              NOT NULL,
    [FixedCoupon]      DECIMAL (19, 14) NOT NULL,
    [AttachmentPoint]  DECIMAL (19, 14) NOT NULL,
    [DetachmentPoint]  DECIMAL (19, 14) NOT NULL,
    CONSTRAINT [PK_Tranched_InstrumentId]   PRIMARY KEY CLUSTERED ([InstrumentId] ASC),
    CONSTRAINT [FK_Tranched_Instrument]     FOREIGN KEY ([InstrumentId]) REFERENCES [instrument].[Instrument] ([InstrumentId]),
    CONSTRAINT [FK_Tranched_Series] FOREIGN KEY ([SeriesId]) REFERENCES [instrument].[Series] ([InstrumentId])
);
```

### Steps to reproduce the problem

The code that creates the merge query is used to populate other tables which also contain decimal(19, 14) columns and it works as expected.

So providing a MCVE may prove a little complicated, but I can try if the information provided is insufficient.

### jOOQ Version

jOOQ Pro 3.19.11

### Database product and version

SQL Server 17

### Java Version

21

### JDBC / R2DBC driver name and version (include name if unofficial driver)

com.microsoft.sqlserver:mssql-jdbc:12.8.1.jre11

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'll investigate this soon.

**Comment by lukaseder** [comment]
Thanks for your patience.

It doesn't seem like this has anything to do with the `MERGE` statement, but just with your `VALUES` constructor usage. This also reproduces it:

```java
public void testValuesConstructorWithBigDecimal() throws Exception {
    bindAndInline(ctx -> {
        BigDecimal b1 = BigDecimal.ONE;
        BigDecimal b2 = new BigDecimal("1.1");
        BigDecimal b3 = new BigDecimal("1.10");

        Result<Record2<Integer, BigDecimal>> r1 = ctx.selectFrom(values(
            row(1, b1),
            row(2, b2),
            row(3, b3)
        )).fetch();

        Result<Record2<Integer, BigDecimal>> r2 = ctx.selectFrom(values(
            row(val(1), val(b1, DECIMAL(3, 2))),
            row(val(2), val(b2, DECIMAL(3, 2))),
            row(val(3), val(b3, DECIMAL(3, 2)))
        )).fetch();

        assertEquals(asList(1, 2, 3), r1.map(r -> r.value1()));
        assertEquals(b1, r1.get(0).value2());
        assertEquals(b2, r1.get(1).value2());
        assertEquals(b3, r1.get(2).value2());
        assertEquals(asList(1, 2, 3), r2.map(r -> r.value1()));
        assertEquals(b1, r2.get(0).value2());
        assertEquals(b2, r2.get(1).value2());
        assertEquals(b3, r2.get(2).value2());
    });
}
```

When run with bind variables, the statement produces this result:

```
select [v].[c1], [v].[c2] from (values (?, ?), (?, ?), (?, ?)) [v] ([c1], [c2])
+----+----+                                                                    
|  c1|  c2|                                                                    
+----+----+                                                                    
|   1|   1|                                                                    
|   2|   1|                                                                    
|   3|   1|                                                                    
+----+----+                                                                    
```

This helps work around the problem:

```java
BigDecimal b1 = BigDecimal.ONE.setScale(2);                                                                 
```

Alternatively, individual bind values could be cast explicitly to their data type (`DECIMAL(19, 14)`, for example, using `DSL.cast(DSL.val(number), TRANCHED.FIXEDCOUPON)`.

This reminds me of a previous issue:

- https://github.com/jOOQ/jOOQ/issues/8964

I'll check if this is a regression or a subtly different problem (e.g. the other issue was about the `INSERT` statement only, this seems more general)

**Comment by assylias** [comment]
Hi Lukas, thanks for your answer.

The source data comes from a `DSL.values(rows)` where rows is an `Array<RowN>` (kotlin).

Extract from the code below:

    val records = -...

    val source = SourceTable(table, records, emptySet())

    val targetInstId = table.field(INSTRUMENT_.INSTRUMENTID.name, SQLDataType.INTEGER)!!
    val sourceInstId = source.values.field(INSTRUMENT_.INSTRUMENTID.name, SQLDataType.INTEGER)!!

    dsl().mergeInto(table)
      .using(source.values)
      .on(sourceInstId.eq(targetInstId))
      .whenMatchedThenUpdate()
      .set(source.fields.associateWith { source.field(it) })
      .whenNotMatchedThenInsert(*source.fields)
      .values(*source.values.fields())
      .execute()

And the `SourceTable` class is:

```
data class SourceTable<R: UpdatableRecord<R>, T: Table<R>>(
  val table: T,
  val records: Collection<R>,
  val excludedFields: Set<Field<*>> = emptySet()
) {
  private val excludedFieldNames = excludedFields.map { it.name }.toSet()
  val rows = recordsToRows(records, excludedFieldNames)

  val fields = table.fields().filterNot { it.name in excludedFieldNames }.toTypedArray()
  private val fieldNames = fields.map { it.name }.toTypedArray()
  val values = values(*rows).`as`("input", *fieldNames)

  /**
   * @return the given field in the values table created in this class.
   */
  fun <T> field(f: Field<T>) = values.field(f.unqualifiedName, f.type)!!

  /**
   * @return an array of rows containing one row per record, and one column for each included field
   * @param records the records to turn into an array of rows
   * @param excludedFields the fields that should not be included in the rows
   */
  private fun recordsToRows(records: Collection<R>, excludedFields: Set<String> = emptySet()): Array<RowN> {
    if (records.isEmpty()) return emptyArray()
    val f = records.first().fieldsRow()
    val includeIndices = ( 0 until f.size()).filter { i -> f.field(i)!!.name !in excludedFields }

    return records.map {
      val selectedFields = it.valuesRow().fields().sliceArray(includeIndices)
      row(*selectedFields)
    }.toTypedArray()
  }
}
```

**Comment by lukaseder** [comment]
> Note that this is a little more complex than it should because of #17202, but even if that other issue is resolved, I need to exclude some fields from the Set/Insert so I think I still have to do something like this.

Plus, fixing #17202 wouldn't address this issue here, which is independent, in my opinion. The simplest generic workaround that I can think of is to implement a custom `Binding<BigDecimal, BigDecimal>` that always generates the `CAST` for these columns, knowing that this will affect *all* bindings of this data type.

Alternatively, you could work with a temporary table, of course.

**Comment by lukaseder** [comment]
This is merely a type inference problem when all types are unknown and inferred from bind values. E.g. this works:

```sql
select [v].[c1], [v].[c2]
from (
  values
    (1, cast(1 as numeric(1))),
    (2, cast(1.1 as numeric(2, 1))),
    (3, cast(1.10 as numeric(3, 2)))
) [v] ([c1], [c2])
```

So, the fix #822 that we've applied already for numerous other dialects can apply here as well.

There's a bit of subtlety hidden in the approach taken by this fix here:

- https://github.com/jOOQ/jOOQ/issues/8964

It uses the type of the `INSERT` statement's target column rather than the scale of the bind variable, which is obviously better, if it is known. Specifically, our test cases for #8964 test bulk and batch insertion. In the latter case, it's important that the first batch produces a SQL string with SQL `CAST` expressions that won't truncate decimal values for subsequent batches. If you're batching your merge statements, then this can be a problem in this case as well.

I've created a separate issue for this:

- https://github.com/jOOQ/jOOQ/issues/17414
